File: xiaqianZhang/assignments/treesDFS/lc112/lc112.py
# Problem Statement #

# Given a binary tree and a number ‘S’, 
# Find if the tree has a path from root-to-leaf such that the sum of all the node values of that path equals ‘S’.
'''
Attempt 1:
Input: root of the binary tree and the target sum that we want to find
Output: boolean value to check if there's root to leaf path where the sum match the target sum
Edge case: if the root is none: return false

check if the current node is not a leaf node, if so continue looping till its a leaf node, 
  by checking if the current node have left and right children, if so change the current node to its children
    at the same time, we want to subtract the current node's value from the sum sum = sum - current node's value

create a recrusive function that repeatly checking the following  def rec(root, sum)
  if (the root.left is null and root.right is null)[it is a leaf node]:
    sum -= current node's value
    if sum == 0:
      return true
    else:
      return false
run the rec function by either take the input replace the root as root.left or root.right

'''

class TreeNode:
  def __init__(self, val, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

# Given a binary tree and a number ‘S’, 
# Find if the tree has a path from root-to-leaf such that the sum of all the node values of that path equals ‘S’.
'''

'''

def has_path(root, sum):
  
  def rec(root, sum):
    if root is None:
      return False

    sum -= root.val
    
    if root.left is None and root.right is None:
      if sum == 0:
        return True
      else:
        return False
    
    if(rec(root.left, sum) or rec(root.right, sum)):
      return True
    return False

  return rec(root, sum)

    

def main():

  root = TreeNode(12)
  root.left = TreeNode(7)
  root.right = TreeNode(1)
  root.left.left = TreeNode(9)
  root.right.left = TreeNode(10)
  root.right.right = TreeNode(5)
  print("Tree has path: " + str(has_path(root, 23))) #true 12 + 1 + 10
  print("Tree has path: " + str(has_path(root, 16))) #false 

  root1 = None
  print("Tree has path: " + str(has_path(root1, 16))) 


main()

File: henryPuma/assignments/treesDFS/lc112/lc112.js
// Problem Statement #

// Given a binary tree and a number ‘S’, find if the tree has a path from root-to-leaf such that the sum of all the node values of that path equals ‘S’.

class TreeNode {

  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
};

/*
Precondition: 
root: The root of the binary tree, if null automatically return false
sum: The sum we want from the root to the leaf

Question: Are all the nodes going to have values < sum? Yes


Post condition: 
Return a boolean that tells us whether or not a sum from root->leaf exists and === sum

nodeStack: Keeps track of all the future nodes that we are going to look at,
sumStack: Keep the compliment sum of the current node and the targetSum <=> sum (parameter)

LEAF NODE: A node where node !== null, node.left and node.right === null


General Idea:
Keep a nodeStack list that keeps track of the nodes being processed,
Keep a sumStack list that keeps track of the overall sum of the nodes that have been processed,
Have 2 variables: currentNode and currentSum that temporarily hold the popped values from nodeStack and sumStack,
currentNode: Takes the next node for processing,
currentSum: Takes the overall sum of the nodes processed,

LINE: 52 Check if the current node is a leaf, if it is ==> check if the currentSum === 0 meaning we've subtracted node values that add up to sum ("Target sum")
If this case is met then return TRUE

Otherwise, check the currentNode's children (left && right),
If either exist, then add that node to the nodeStack and update the sumStack with currentSum where (currentSum - currentNode.child.value) gets pushed,


If for some reason we process all the paths in the binary tree and NO PATH adds up to sum, then by default we return FALSE

*/

const has_path = function(root, sum) {
  if(!root) return false;

  const nodeStack = [root];
  const sumStack = [sum - root.value];

  let currentNode = null, currentSum = 0;

  while(nodeStack.length){ // Iterate at most N times. In the worst case we exhaust all paths from root to leaves.
    currentNode = nodeStack.pop(); 
    currentSum = sumStack.pop(); 
    if (!currentNode.left && !currentNode.right && currentSum === 0) return true;

    if (currentNode.left){
      nodeStack.push(currentNode.left); 
      sumStack.push(currentSum - currentNode.left.value);
    }

    if (currentNode.right){
      nodeStack.push(currentNode.right); 
      sumStack.push(currentSum - currentNode.right.value);
    }
  }
  // Runtime: O(n) where n: number of nodes in the binary tree
  // Space: O(n) where n: number of nodes in the binary tree
  return false;
};
/*
    12 
  7    1
9     10  5

*/

var root = new TreeNode(12)
root.left = new TreeNode(7)
root.right = new TreeNode(1)
root.left.left = new TreeNode(9)
root.right.left = new TreeNode(10)
root.right.right = new TreeNode(5)
console.log(`Tree has path: ${has_path(root, 23)}`) // True, Output: True
console.log(`Tree has path: ${has_path(root, 16)}`) // False, Output: False
